scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt

```diff
@@ -115,33 +120,32 @@ class ClearlyDefinedStorage(
                 .groupBy({ it.substringBefore('/') }, { it.substringAfter('/') })
                 .mapValues { (_, versions) -> versions.sortedWith(AlphaNumericComparator) }
 
-            toolVersionsByName.mapNotNull { (name, versions) ->
-                // See https://github.com/clearlydefined/service#tool-name-registry.
-                when (name) {
-                    "scancode" -> {
-                        val version = versions.last()
-                        val data = loadToolData(coordinates, name, version)
-
-                        data["content"]?.let { result ->
-                            val summary = generateSummary(result)
-                            val details = getScanCodeDetails(result)
-
-                            val provenance = getProvenance(coordinates)
+            val supportedScanners = toolVersionsByName.mapNotNull { (name, versions) ->
+                // For the ClearlyDefined tool names see https://github.com/clearlydefined/service#tool-name-registry.
+                ScannerWrapper.ALL[name]?.let { factory -> factory to versions.last() }.also {
+                    if (it == null) logger.debug { "Unsupported tool '$name' for coordinates '$coordinates'." }
+                }
+            }
 
-                            ScanResult(provenance, details, summary)
-                        } ?: run {
-                            logger.warn {
-                                "'$coordinates' has no data available for tool '$name' in version '$version'."
+            supportedScanners.mapNotNull { (factory, version) ->
+                val scanner = factory.create(ScannerConfiguration(), DownloaderConfiguration())
+                (scanner as? CommandLinePathScannerWrapper)?.let {
+                    val startTime = Instant.now()
+                    val name = factory.type.lowercase()
+                    val data = loadToolData(coordinates, name, version)
+
+                    when (factory.type) {
+                        "ScanCode" -> {
+                            data["content"]?.let { result ->
+                                val provenance = getProvenance(coordinates)
+                                val details = getScanCodeDetails(result)
+                                val endTime = Instant.now()
+
+                                ScanResult(provenance, details, it.createSummary(result.toString(), startTime, endTime))
```

Comment:
> I assumed extending the `CommandLinePathScannerWrapper` API in preceding commit would allow for injecting handling for new scanners without the need for making code modification.

That was my plan as well... until I realized that the JSON response from ClearlyDefined is tool-specific.

> With above abstraction, it should be possible to fix the issue of tool specific code in here.
What do you think?

I think it's a good idea, and it would probably pay off even more if ORT and ClearlyDefined had more than two tools in common, but I'd like to keep such a refactoring out of this PR to not make it grow too large.

Reply:
> I think it's a good idea

If we'd agree on this idea (probably needs to be detailed out a bit / synced)

> but I'd like to keep such a refactoring out of this PR to not make it grow too large

...wouldn't this refactoring move into a direction further away from the above? e.g. the extension of that scanner interface, the code which I flagged which adds child class specific logic, also the code locations. E.g. scancode specific code could probably remain in the `scancode` package, perspectively being extracted to a scancode plugin. 
The logic for parsing clearly defined results could then be called via plugin mechanism, instead of having it located outside of the scanner. As it is closely related to scancode, the logic could remain in scancode plugin, so the scancode plugin would just implement `ScannerFactory` as well as `ClearyDefineResultParserFactory`. Then all scanner plugins could optionally implement any of these two interfaces.

Reply:
> ...wouldn't this refactoring move into a direction further away from the above?

I don't think so. The current (maybe now intermediate) refactoring nicely groups the code that should be abstracted away via the `ClearlyDefineResultParser` interface. I'd keep that interface in `ClearlyDefinedStorage.kt`, probably.

Reply:
I'm ok if this is just an intermediate step, in particular as it helps to move scancode scanner to it's own plugin dir. I'll approve it. Would be nice to have a second approval though, as I'm not really sure it isn't controversial.